### PR TITLE
[WIP] Add support for drag & drop re-ordering of submenus

### LINKED ISSUE #4: Drag & Drop Re-ordering needs to account for submenus
Keyboard-accessible reordering needs to account for moving parent items as well, to move the submenu items with it.

### PR DESCRIPTION
**Pull request by westonruter**
Fixes #4.

### CODE REVIEW

**Comment by GaryJones** [comment]
I realise a WIP, but some feedback:

* Sub-sub (grandchild) menu items are buggy. Moving them around makes them appear as child items in the preview (Utility Pro child theme for Genesis, if that makes any difference), even if the Customizer still shows it as being a grandchild item.
* With the drag handle being all the way across, it's very easy to grab too far to the right of an item, and then not have sufficient space on the left to move a child / grandchild item up a level. Suggest that the drop target look at the top left of the menu item, and not where the cursor is. Top right if the Customizer is shown on the right for RTL.
* _sub item_ marker isn't really scaleable - grandchild items may be expected to be _sub sub item_, and so on. Visually, the structure can already be seen, and for accesibility, I suggest adding aria attributes / `wp.ally.speak()` that says something like "Child Cat 3.1 is a child item of Parent Cat 3".

**Comment by valendesigns** [comment]
Not sure why all of the JS issues that I fixed in the master will not merged into this branch, but I guess I'll fix it again real quick so the build passes.

**Comment by adamsilverstein** [comment]
Thanks Derek. There were a bunch of merge conflicts I didn't have a chance
to work out.

On Saturday, June 6, 2015, Derek Herman <notifications@github.com> wrote:

> Not sure why all of the JS issues that I fixed in the master will not
> merged into this branch, but I guess I'll fix it again real quick so the
> build passes.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/voldemortensen/menu-customizer/pull/63#issuecomment-109677071>
> .
>


-- 
Adam Silverstein
Engineering Manager @ 10up

**Comment by valendesigns** [comment]
No worries, I'll get it fixed in a commit or two.

**Comment by valendesigns** [comment]
And fixed!

**Comment by adamsilverstein** [comment]
Great! Almost to the bottom of the weird submenu positioning issues.

**Comment by valendesigns** [comment]
What are the outstanding issue?

**Comment by valendesigns** [comment]
What I found in 2 minutes.
* It looks like you can make a top level a sub menu of nothing.
* Preview doesn't work.
* The live Menu doesn't use the saved order.

**Comment by adamsilverstein** [comment]
Derek,

Can you test with the latest commit. I may have some things disabled while
I iron out the submenu placement glitches, This should be resolved:

   - It looks like you can make a top level a sub menu of nothing.

Let me check the other issues!

On Sat, Jun 6, 2015 at 9:20 PM, Derek Herman <notifications@github.com>
wrote:

> What I found in 2 minutes.
>
>    - It looks like you can make a top level a sub menu of nothing.
>    - Preview doesn't work.
>    - The live Menu doesn't use the saved order.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/voldemortensen/menu-customizer/pull/63#issuecomment-109680103>
> .
>



-- 
Adam Silverstein
Engineering Manager @ 10up

**Comment by adamsilverstein** [comment]
Still seeing issues. Hmph! Can you try testing with only one submenu depth - some issues may be specific to 'grandchildren'

**Comment by valendesigns** [comment]
Pulled and is still an issue with the first top level menu item. Is there code from reorder that we can use here, or have we been doubling our efforts?

**Comment by valendesigns** [comment]
This branch cannot be merged the way it is currently. Not because it's not finished, because it's deleting code that needs to stay in the master branch. When the conflicts were not dealt with properly the branch got confused about keeping code that was recently added. I'm not sure what the best approach is here, but we cannot merge as is.

**Comment by westonruter** [comment]
> because it's deleting code that needs to stay in the master branch

Which code?

**Comment by adamsilverstein** [comment]
I'm not sure about the reorder work, I haven't been working on that part. I recently merged master into this branch and didn't have much trouble. I have a fairly clean diff against master: https://cloudup.com/cOEGn1eOk5T

**Comment by valendesigns** [comment]
The changes in `setupUIPreviewing` are what stands out the most. I see other small issues here and there that either add or remove things incorrectly like a `(` and `)` that were removed from a conditional statement, but are now added back in. Somethings are minor, but some are not.

File: menu-customizer.js [hunk 1]
```diff
@@ -1143,6 +1165,7 @@
 				inputs = $( self.container ).find( ':input[name]' );
 				inputs.each( function() {
 					var name = this.name;
+					name = name.replace( /\[\d+]/, '' ); // Remove the ID-part of the name which is used by nav-menus.php.
```
Comment (valendesigns) [inline menu-customizer.js]: What happens when the menu name has a number in it, will this regex still work as expected?

File: menu-customizer.js [hunk 2]
```diff
@@ -1608,150 +1641,67 @@
...
+		 *
+		 * @param {object} menuList - The element that has sortable().
 		 */
-		_setupSortable: function() {
+		_setupSortable: function( menuList ) {
 			var self = this;
 
-			this.isReordering = false;
+			if ( ! menuList.is( self.$sectionContent ) ) {
+				throw new Error( 'Unexpected menuList.' );
```
Comment (valendesigns) [inline menu-customizer.js]: This string needs to be added to the `l10n` object.

File: menu-customizer.js [hunk 3]
```diff
@@ -1999,7 +1953,7 @@
 				event.stopPropagation();
 				event.preventDefault();
 			} );
-			$( '#accordion-panel-menus' ).on( 'click keydown', '.menu-delete', function( event ) {
+			$( '#accordion-panel-menus' ).on( 'click keydown', '.menu-delete', function() {
```
Comment (valendesigns) [inline menu-customizer.js]: `event` is missing.

File: menu-customizer.js [hunk 4]
```diff
@@ -2296,13 +2239,12 @@
 	 */
 	function setupUIPreviewing() {
 		$( '#accordion-panel-menus' ).on( 'input', '.live-update-section-title', function( e ) {
-			var input = $( e.currentTarget ),
```
Comment (valendesigns) [inline menu-customizer.js]: The fix to update locations in checkboxes has been reverted for some reason.

Comment (adamsilverstein) [inline menu-customizer.js]: Likely from a failed merge. I can start back at master and redo the changes
to clean this up.

On Saturday, June 6, 2015, Derek Herman <notifications@github.com> wrote:

> In menu-customizer.js
> <https://github.com/voldemortensen/menu-customizer/pull/63#discussion_r31874081>
> :
>
> > @@ -2296,13 +2239,12 @@
> >  	 */
> >  	function setupUIPreviewing() {
> >  		$( '#accordion-panel-menus' ).on( 'input', '.live-update-section-title', function( e ) {
> > -			var input = $( e.currentTarget ),
>
> The fix to update locations in checkboxes has been reverted for some
> reason.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/voldemortensen/menu-customizer/pull/63/files#r31874081>
> .
>


-- 
Adam Silverstein
Engineering Manager @ 10up

**Comment by adamsilverstein** [comment]
@valendesigns To avoid any further merge issues, I started at master and cherry picked all the changes from this branch to a new branch, here: https://github.com/voldemortensen/menu-customizer/tree/feature/sub-menus-2

I hope that helps,  the submenus are working again locally although still seeing the same remaining issues.

**Comment by westonruter** [comment]
Amended patch to my Core PR https://github.com/voldemortensen/menu-customizer/pull/73: https://github.com/xwp/wordpress-develop/commit/0f89312d238aa7ef80316d6d77633eaacf1e243f

**Comment by valendesigns** [comment]
@adamsilverstein I'm going to pick up on this issue in a couple minutes. I'll open a new PR with the alternative branch. Thank you for creating a clean version! +1

**Comment by valendesigns** [comment]
Closing in favor of #99

**Pull request closed without merging**